### DISCUSSION TITLE
Can Dreambooth read accompanying text files with descriptions/captions?

### DISCUSSION THREAD

**Discussion by wktra** [d8ahazard/sd_dreambooth_extension#399] [Q&A]
Sorry to beat this dead horse, but it seems like this question is like asking if God is real. 😓 No one on several discords can answer this question.

I've tried many settings with no luck. Or maybe it's that I just have no idea if it's working with DB. 

Does anyone know the answer? 

**Can Dreambooth read accompanying text files with descriptions/captions for training images?**

**Comment by D4ndellion** [comment]
yes.

**Reply by wktra** [reply]
Thank you 🤧

**Reply by wktra** [reply]
> yes.
@dali99 
okay, so I appreciate your short answer, but dreambooth is not reading my text files. I get this error:

```
   File "Q:\AI\Super_SD_2.0\stable-diffusion-webui\venv\lib\site-packages\PIL\Image.py", line 2317, in save
    fp = builtins.open(filename, "w+b")
OSError: [Errno 22] Invalid argument: 'Q:\\AI\\Super_SD_2.0\\stable-diffusion-webui\\models\\dreambooth\\773sloppy\\samples\\773sloppy 1girl bangs blunt bangs blush breasts cleavage cleavage cutout clothing cutout demon girl demon horns demon tail demon wings eyebrows visible through hair full body heart cutout horns looking at viewer medium breasts nail polish navel open mouth pink eyes pink hair pink nails pointy ears simple background solo stirrup legwear tail toeless legwear toenail polish toenails toes twintails wings double hair bun500-0.png'
[*] Weights saved at Q:\AI\Super_SD_2.0\stable-diffusion-webui\models\dreambooth\773sloppy
```

exactly what am I doing wrong?????

**Reply by D4ndellion** [reply]
Can't help you I don't use windows

**Reply by wktra** [reply]
> Can't help you I don't use windows

But I use windows and I'm begging you for help. 
I'm literally **BEGGING** you for help. 
If you just look and try to guess at what the problem might be, I'll pass it on to ANYONE else who might have this problem.
PLEASE. PLEASE @dali99

**Reply by D4ndellion** [reply]
It's not like I'm withholding knowledge because you're using windows. I just don't know how your system works...

Cursory googling suggests it's not parsing the path you've provided as a path, why this happens with the string above is something you'll have to reason through. I don't know what kinds of things are permissible or need escaping  in the insanity that is windows platform filepaths.

try make it shorter maybe. I'd recommend using  simpler filenames and putting your tags in accompanying text files rather than this approach you seem to be using.

The begging isn't really how we do things on these platforms. Post your question, refrain from tagging and notifying people directly. Wait for someone who does know the answer to come along and answer it. Your comments should be mainly adding more information or documenting things you've tried.

Again I can't help you and I can't dedicate my time to holding your hands through troubleshooting. So aiming your question at the general public rather than me, filing a bug report with the error you're getting, is a much more effective way of getting an answer :)

**Reply by wktra** [reply]
> It's not like I'm withholding knowledge because you're using windows. I just don't know how your system works...
> 
> Cursory googling suggests it's not parsing the path you've provided as a path, why this happens with the string above is something you'll have to reason through. I don't know what kinds of things are permissible or need escaping  in the insanity that is windows platform filepaths.
> 
> try make it shorter maybe. I'd recommend using  simpler filenames and putting your tags in accompanying text files rather than this approach you seem to be using.
> 
> The begging isn't really how we do things on these platforms. Post your question, refrain from tagging and notifying people directly. Wait for someone who does know the answer to come along and answer it. Your comments should be mainly adding more information or documenting things you've tried.
> 
> Again I can't help you and I can't dedicate my time to holding your hands through troubleshooting. So aiming your question at the general public rather than me, filing a bug report with the error you're getting, is a much more effective way of getting an answer :)

I appreciate your response, but please know this:

I AM using accompanying text files. It's pulling those tags FROM accompanying text files!

The training images are all in this fashion: 773sloppy (1).png, 773sloppy (1).txt, 773aloppy (2).png, 773sloppy (2).txt, etc etc

**Reply by exdownloader** [reply]
You might have a file path length issue. I recently made some changes (locally) to the preview saving to avoid a file path bug, where the sanitised prompt would trigger a failure to save previews. To that end, try renaming your files such that the full file path is below 256 characters. So long as there is an accompanying text file with the same file name (excluding extension), it'll read fine, as you've listed.

**Reply by wktra** [reply]
@exdownloader you've done me a great kindness. Thank you. I'm going to shorten the path see where that leads me. 🙏🤧💖

**Comment by cvang187** [comment]
Sidenote: It seems you need to enable filepath beyond the default char limit that Windows has by default. https://www.howtogeek.com/266621/how-to-make-windows-10-accept-file-paths-over-260-characters/

This is how I auto caption each instance image and setup the dreambooth extension to use the concepts_list.json file located in sd_dreambooth_extension\dreambooth:

1) use the AutoCaption.ipynb tool from [EveryDream-Trainer](https://github.com/victorchall/everydream).  Most likely you'll need a local install of Jupyter Notebook or just copy the files to Google Colab as it runs juypter-notebooks. There are other alternatives but I find this one works quite well.

2) Get Windows PowerToy from the Microsoft Store. It has two very useful feature called "PowerRename" and also one called "ImageResizer." ImageResizer allows you to batch/mass resize images to any demension you want thus saving you time using online services or other programs.

3) Use PowerRename after you've auto-captioned each image with EveryDream-Trainer's juypternotebook. PowerRename allows you to select a bunch of files and rename/append/find and replace and use RegularExpression to batch modify any file name of any file type.  You need it to add your instance token and class token to the begining of the image file name. In the example below, "opyamato" is the instance token and "woman" is the class token. Following them is the autocaptioned image description from EveryDream-Trainer.

![Screenshot_20221204_033214](https://user-images.githubusercontent.com/8205798/205516779-11d9bec0-a523-40e7-8798-9c67db32fb28.png)

4) If your images are below 512x512 image size, I'd suggest you use the upscalers in Auto1111's web-ui to upscale them by 2x or so and then use the Window's PowerTool - ImageResize to select all your instance image and resize them to 512x512

5) configure your concepts_list.json to something similar to my example below. This is an excerpt from my concept_list.json that is currently setup to train for 11 characters. You can add each { } within the json file for any x amount of character/concept you would like to train dreambooth with. It is important to note, you need to make sure "file_prompt_contents": "Instance + Class + Description", is set for it to read the instance. class, and description from your image file name. I also pasted my full concepts_list.json to paste bin since it doesnt seem like github is playing nicely with the json code snippet below. [concepts_list.json](https://pastebin.com/nv0nf7jd)

`[
    {
        "max_steps": -1,
        "instance_data_dir": "D:\\onepiece\\op_woman\\nico",
        "class_data_dir": "D:\\class_image_w",
        "file_prompt_contents": "Instance + Class + Description",
        "instance_prompt": "[filewords]",
        "class_prompt": "[filewords]",
        "save_sample_prompt": "opnico",
        "instance_token": "opnico",
        "class_token": "woman",
        "num_class_images": 2000,
        "class_negative_prompt": "",
        "class_guidance_scale": 7.5,
        "class_infer_steps": 40,
        "save_sample_negative_prompt": "",
        "n_save_sample": 1,
        "sample_seed": -1,
        "save_guidance_scale": 7.5,
        "save_infer_steps": 40
    },
    {
        "max_steps": -1,
        "instance_data_dir": "D:\\onepiece\\op_woman\\yamato",
        "class_data_dir": "D:\\class_image_w",
        "file_prompt_contents": "Instance + Class + Description",
        "instance_prompt": "[filewords]",
        "class_prompt": "[filewords]",
        "save_sample_prompt": "opyamato",
        "instance_token": "opyamato",
        "class_token": "a woman",
        "num_class_images": 2000,
        "class_negative_prompt": "",
        "class_guidance_scale": 7.5,
        "class_infer_steps": 40,
        "save_sample_negative_prompt": "",
        "n_save_sample": 1,
        "sample_seed": -1,
        "save_guidance_scale": 7.5,
        "save_infer_steps": 40
    }
]`

**Reply by wktra** [reply]
@sayoonarachu 
I appreciate the advice.

So basically, you're telling me not to use an accompanying text file with the descriptions. 

You're telling me to just rename the file with the captions.

In which case, this Dreambooth cannot read captions/descriptions from accompanying text files?

**Reply by FugueSegue** [reply]
Yes, you can have Dreambooth read instance dataset text prompt files. I have started using this technique with the best results I've ever had with training a model, embedding or hypernetwork.

The name of the text file must be the same as the image file. Perhaps the trouble you're running into is the parenthesis characters. Or maybe the space characters. I don't know for sure. I always use letters, numbers, and underscore characters for all my filenames. This is a habit of mine that dates back to toiling with MS-DOS in the '80s.

What I can't answer for certain is the best way to properly prepare and use them. I've been feeding my instance images into Automatic1111's BLIP caption generator in the Preprocess Images tab under the Train tab. Normally this is used for preparing embeddings or hypernetworks. But it works for this task as well. Using this tool, it automatically renames your images files and creates text prompt files with the same names.

The next part I'm not entirely certain is correct. And perhaps how all this is done will change in the near future. But this is my understanding of it at this point.

For Dreambooth to properly interpret the text prompt files, you must edit each one of them. Replace the words given for your subject with the Instance Token and the Class Token you specify under the Filewords section in the Concepts tab. For example, if you were training a model to learn the face of a woman named Jane, your instance token could be anything you want such as "jane1234" or "myxlplyx67"--whatever you want. But I would highly recommend that you use "person" as your Class Token. Not "woman". I'm not sure but I think "person" works better. After you feed your instance dataset images through BLIP, one of the text files it generates might read:

a woman with black hair wearing a red shirt

You would need to edit this to read:

jane1234 person with black hair wearing a red shirt

It's also possible that BLIP will make mistakes. Such as specifying that Jane is carrying a gun when she's actually carrying a book. Correct these mistakes.

All of this text file editing is tedious but it pays off in quality.

Again, I'm certainly not an expert and I am still trying to learn about this stuff, the same as you. I am also very frustrated when I seek help on the internet, only to receive vague, one-word answers without context or examples.

**Reply by wktra** [reply]
I highly appreciate your goodwill and advice. I will take it and if it serves me well, I will update this and pass it on! Thank you!

**Reply by wktra** [reply]
@GunnarQuist I took your advice and made all of those suggested changes.

The key to fixing the error was to include a concept list with my one subject. The only difference between how I listed the path to the dataset in the concept list _vs_ webui is that I actually had to type double-backslashes for paths in the json.  Apparently, the DB does weird stuff with filewords and just entering the path normally in the webui. Putting this into the concept list json fixes this. (but keeping in mind that paths in the json need double backslashes)

Thank you for the advice. 

One question: When training this way, I get great and accurate results, BUT...
**I have to type full descriptive captions to generate that trained character.** It's not enough to just type the instance token and class prompt. Is this normal?

**Reply by D4ndellion** [reply]
any words which you don't have examples without or being different will essentially become the instance prompt. Regularization can help

**Reply by wktra** [reply]
@dali99 And by regularization, you mean I would need regularization class images WITH description/captions OF THOSE EXAMPLES, correct?

This is for clarification in case people refer back to these posts  and need help with filewords and training.

**Reply by D4ndellion** [reply]
yes, the idea is to help it separate the concepts from each other

**Reply by FugueSegue** [reply]
TLDR: If you specify "[filewords]" as the string for the class prompt, does it then read text prompt files that correspond to each and every classification image? The Dreambooth extension doesn't generate text prompt files when it automatically generates classification images based on a prompt you can specify. 

Text prompt files for each classification image? That's a lot! Sometimes I have over 3,000 classification images for one training.

Up to this point I've been training only one concept per model training. So I haven't bothered using JSON files yet. I've only used the webui to enter parameters for one concept. It automatically creates a JSON file when it begins training.

The default way of generating classification images is to specify a prompt phrase in the Class Prompt text box under the Concepts tab. Then it generates images based on that prompt and puts them in the directory that you specify in the Classification Dataset Directory text box.

I have assumed that if you enter a prompt in the Class Prompt text box that the Dreambooth extension subsequently uses that same prompt with each of the Classification images during training. Do I really have to have a text prompt file for each of the thousands of classification images? If that is the case then it seems terribly redundant. Even if you forego the use of the Concepts tab in favor of a JSON file, you still have to specify a text string for the class_prompt element of the concepts_list, right?

I can imagine that having text prompt files for each classification image might be somehow useful. I think I could manage the automated production of hundreds of classification text prompt files. But doing all of that work would depend on whether or not entering "[filewords]" as the string for the class_prompt element would invoke this course of action.

**Reply by d8ahazard** [reply]
If you use [filewords] as a string for the class prompt, two things happen.

One - any existing class images with prompts in their filename or an accompanying txt file should have the prompts parsed and used when training.

Two - Any new class images generated will use existing captions from your instance images. The class token and instance token allow us to generate the proper prompt format regardless of what is found where. So, your instance images can be captioned with the subject, or the class, or both, and as long as you set "file prompt contents" properly, instance prompts will have the subject + class, and class prompts will have the subject removed if necessary, and class injected if not there. 



If you do not use [filewords] in your class prompt, then whatever is in the prompt box will be used to generate all class images.

**Reply by piyarsquare** [reply]
In instance prompt, do I enter just **[filewords]** or **xyz style [filewords]** if my accompanying text files are Description only? Thank you.

**Reply by FollowMarcos** [reply]
https://www.reddit.com/r/StableDiffusion/comments/zcr644/make_better_dreambooth_style_models_by_using/

**Reply by gitreal2022** [reply]
> yes, the idea is to help it separate the concepts from each other

I think regs just retrains the concept with the given set of images and it will create a very limited representation of those re-trained original classes. In effect, it will mostly ruin them.
